Implement subset portion of `channel_selection` input for `combine_echodata`

b-reyes opened the pull request:
This PR partially addresses issue #819 by allowing users to select a subset of the channels for the `EchoData` objects before combination. The portion of issue #819 allowing for the expansion of channels will be completed at a later time. 

The outlines provided by the comments https://github.com/OSOceanAcoustics/echopype/issues/819#issuecomment-1329545424 and https://github.com/OSOceanAcoustics/echopype/issues/819#issuecomment-1329887541 were used to construct this PR.  In order to complete this implementation a consistency check of the channels needed to be done. This PR also adds unit tests for all functions needed to complete this consistency check. Additionally, a test placeholder for testing `combine_echodata` with the input `channel_selection` was added.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -89,6 +90,56 @@ def check_zarr_path(
+            raise TypeError("Each key of channel_selection must be a string!")
+
+        # make sure all keys are of the form Sonar/Beam_group
+        are_keys_right_form = ["Sonar/Beam_group" in elem for elem in channel_selection.keys()]
```
Seems like we should limit the string to something like`"Sonar/Beam_group(\d{1})"` and use `re.match()`?

b-reyes replied:
Sounds great! I have went ahead and implemented your suggestion.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -154,12 +205,250 @@ def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
+            unique_channels = set(itertools.chain.from_iterable(all_chan_list))
+
+            # raise an error if we have varying channel lengths
+            raise RuntimeError(
+                f"For the EchoData group {ed_group} the channels: {unique_channels}, are "
+                f"not provided in each EchoData object being combined. One can select which "
+                f"channels should be included using the keyword argument channel_selection in "
+                f"combine_echodata."
+            )
```
```suggestion
            raise RuntimeError(
                f"For the EchoData group {ed_group} the channels: {unique_channels} are "
                f"not found in all EchoData objects being combined. Select which "
                f"channels should be included in the combination using the keyword argument "
                f"channel_selection in combine_echodata."
            )
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -154,12 +205,250 @@ def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
+    RuntimeError
+        If ``channel_selection=None`` and all ``channel`` dimensions are not the
+        same across all Datasets.
+    NotImplementedError
+        If ``channel_selection`` is a list and the list of channels are not contained
+        in the ``EchoData`` group for all Datasets
```
```suggestion
    NotImplementedError
        If ``channel_selection`` is a list and the listed channels are not contained
        in the ``EchoData`` group for all Datasets and need to be created and 
        padded with NaN. This "expansion" type of combination has not been implemented.
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -154,12 +205,250 @@ def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
+    ----------
+    sonar_model: str
+        The name of the sonar model corresponding to ``has_chan_dim``
+    has_chan_dim: dict
+        A dictionary created using an ``EchoData`` object whose keys are
+        the groups of the ``EchoData`` object and values specify if that
+        particular group has a ``channel`` dimension
```
```suggestion
    has_chan_dim: dict
        A dictionary created using an ``EchoData`` object whose keys are
        the ``EchoData`` groups and whose values specify if that
        particular group has a ``channel`` dimension
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -154,12 +205,250 @@ def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
+
+    Returns
+    -------
+    dict
```
```suggestion
    channel_selection_dict : dict
```

leewujung replied:
I added this variable name here because it is referred to under one of the Parameters.

Also, what do you think to add a line like "See tests/echodata/test_echodata_combine.py::test_create_channel_selection_dict" for example outputs from this method." maybe as a Note?

b-reyes replied:
> I added this variable name here because it is referred to under one of the Parameters.

Indeed, this variable name should be included, thank you for catching that!

> Also, what do you think to add a line like "See tests/echodata/test_echodata_combine.py::test_create_channel_selection_dict" for example outputs from this method." maybe as a Note?

That sounds like a good note that would be helpful in the future. I will go ahead and add it in the `Notes` section of the docstring as you suggested.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -154,12 +205,250 @@ def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
+    for ed_group, has_chan in has_chan_dim.items():
+
+        # if there are no channel dimensions in the group, set the value to None
+        if has_chan:
+
+            # if the model is not EK80-like set value to the union of the values
+            # of user_channel_selection
+            if sonar_model in ["EK80", "ES80", "EA640"]:
+
+                # if the model is EK80-like
+                if (ed_group in ["Sonar", "Platform", "Vendor_specific"]) or isinstance(
+                    user_channel_selection, list
+                ):
+                    # set value to the union of the values of user_channel_selection
+                    channel_selection_dict[ed_group] = union_beam_chans
+
+                else:
+
+                    # set value to the user provided input with the same key
+                    channel_selection_dict[ed_group] = user_channel_selection[ed_group]
+
+            else:
+                channel_selection_dict[ed_group] = union_beam_chans
```
This logic here seems can be simplified?
The only case when you have `channel_selection_dict[ed_group] = user_channel_selection[ed_group]` is when you are _not_ in the "Sonar", "Platform", "Vendor_specific" groups and when `user_channel_selection` is _not_ a list.
In all other cases `channel_selection_dict[ed_group] = union_beam_chans` is used.

b-reyes replied:
I believe I understand where the simplification should occur. I will go ahead and implement it.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -220,6 +519,15 @@ def combine_echodata(
+    RuntimeError
+        If any ``EchoData`` group has a ``channel`` dimension value
+        with a duplicate value.
+    RuntimeError
+        If ``channel_selection=None`` and all ``channel`` dimensions are not the
+        same across ``echodatas`` with the same group.
+    NotImplementedError
+        If ``channel_selection`` is a list and the list of channels are not contained
+        in the ``EchoData`` group across all values in ``echodatas``.
```
```suggestion
    RuntimeError
        If ``channel_selection=None`` and the ``channel`` dimensions are not the
        same across the same group under each object in ``echodatas``.
    NotImplementedError
        If ``channel_selection`` is a list and the listed channels are not contained
        in the ``EchoData`` group across all objects in ``echodatas``.
```

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -994,8 +962,16 @@ def combine(
+            # collect the group Dataset from all eds that have their channels unselected
+            all_chan_ds_list = [ed[ed_group] for ed in eds if ed_group in ed.group_paths]
+
+            # select only the appropriate channels from each Dataset
+            ds_list = [
+                ds.sel(channel=ed_group_chan_sel[ed_group])
+                if ed_group_chan_sel[ed_group] is not None
+                else ds
+                for ds in all_chan_ds_list
+            ]
```
I like this implementation! 🎉

leewujung reviewed (approved):
@b-reyes : this looks great, thanks! My comments are mostly textual, except for one place I felt that the logic can be simplified (I had not tried to evaluate those to verify, but it should be pretty easy to do with the machinery you have set up?).

I think this is ready to be merged once the suggested changes are incorporated. I agree that the we should implement a few tests with mock EchoData objects of different sonar_model types. I remember that you did prototype using actual files, is that correct?

b-reyes commented:
@leewujung I have implemented the changes you have suggested. For the portion where you think the logic could be simplified, I believe I understood the simplification you wanted. I have went ahead and implemented a simplification of the logic, but please double check this. 

> I agree that the we should implement a few tests with mock EchoData objects of different sonar_model types. I remember that you did prototype using actual files, is that correct?

I did prototype this function using actual files, so it should be fine. I also have tests there for constructing the channel selection dictionary for `EK60-like, EK80-like,AZFP` sensors. This constructed dictionary is really the crux of the implementation, so I believe the code is good for all of these sensors.

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -346,23 +350,16 @@ def _create_channel_selection_dict(
 
             else:
+                # set value to the union of the values of user_channel_selection
                 channel_selection_dict[ed_group] = union_beam_chans
```
Yes this is what I was referring to. The only thing I would suggest change here is to move `not isinstance(user_channel_selection, list)` to the first so that it gets evaluated first, since it is probably more frequent that we deal with list supplied by the user rather than dict.

leewujung reviewed (approved):
The changes look good. Once you move that one check, please feel free to merge.
